gql/comparator.go

```diff
@@ -0,0 +1,207 @@
+	if value.Kind() != reflect.Slice {
+		return nil, errors.New("filter argument 'Values' should be a list of strings")
+	}
+
+	values := make([]string, value.Len())
+	for i := 0; i < value.Len(); i++ {
+		rawItem := value.Index(i).Interface()
+		item, ok := rawItem.(string)
+		if !ok {
+			return nil, errors.New("filter argument 'Values' should be a list of strings")
+		}
+		values[i] = item
+	}
+
+	return inComparator{values: values}, nil
+}
+
+// NewIntegerComparator returns a NewListOperator for integer operations supported by integerComparator,
+// specifically <, <=, > and >=.
+// The returned NewListOperation expects an int with the key "Value" as part of the given argument.
+func NewIntegerComparator(op string) NewListOperation {
+	switch op {
+	case ">", ">=", "<", "<=":
+		break
+	default:
+		return func(arg map[string]interface{}) (Comparator, error) {
+			return nil, fmt.Errorf("unsupported integer comparison operation %q", op)
+		}
+	}
+	return func(arg map[string]interface{}) (Comparator, error) {
+		raw, ok := arg["Value"]
+		if !ok {
+			return nil, errors.New("filter argument is missing 'Value'")
+		}
+		value, ok := raw.(int)
+		if !ok {
+			return nil, errors.New("filter argument 'Value' must be an integer")
+		}
+
+		return integerComparator{operation: op, value: value}, nil
+	}
+}
+
+// NewNotInComparator wraps NewInComparator returning the opposite boolean.
+func NewNotInComparator(arg map[string]interface{}) (Comparator, error) {
+	in, err := NewInComparator(arg)
+	if err != nil {
+		return nil, err
+	}
+	return NotComparator{Child: in}, nil
+}
+
+// NewLimitLengthComparator returns a comparator which simply limits the length of the results to the integer
+// given with the key "Value" in the argument.
+func NewLimitLengthComparator(arg map[string]interface{}) (Comparator, error) {
+	raw, ok := arg["Value"]
+	if !ok {
+		return nil, errors.New("filter argument is missing 'Value'")
+	}
+	limit, ok := raw.(int)
+	if !ok {
+		return nil, errors.New("filter argument 'Value' must be an integer")
+	}
+
+	return &limitLength{limit: limit}, nil
+}
+
+type inComparator struct {
+	values []string
+}
+
+func (c inComparator) Match(raw interface{}) bool {
+	in, ok := raw.(string)
+	if !ok {
+		return false
+	}
+	for _, v := range c.values {
+		if in == v {
+			return true
+		}
+	}
+	return false
+}
+
+type intEqual struct {
+	value int
+}
+
+func (c intEqual) Match(raw interface{}) bool {
+	in, ok := raw.(int)
+	if !ok {
+		return false
+	}
+	return in == c.value
+}
+
+// integerComparator supports comparisons between ints for a these operations <, <=, > and >=.
+type integerComparator struct {
+	operation string
+	value     int
+}
+
+func (c integerComparator) Match(raw interface{}) bool {
+	in, ok := raw.(int)
+	if !ok {
+		return false
+	}
+	switch c.operation {
+	case ">":
+		return in > c.value
+	case ">=":
+		return in >= c.value
+	case "<":
+		return in < c.value
+	case "<=":
+		return in <= c.value
+	default:
+		return false
```

Comment:
Seems more appropriate to `panic` here since this should never happen.

Reply:
Yeah it should be impossible, I'll update